src/pygeom.c

```diff
@@ -78,39 +89,43 @@ static PyObject *GeometryObject_str(GeometryObject *self)
 
 static Py_hash_t GeometryObject_hash(GeometryObject *self)
 {
-    void *context = geos_context[0];
     unsigned char *wkb;
     size_t size;
     Py_hash_t x;
 
     if (self->ptr == NULL) {
         return -1;
     }
-    GEOSWKBWriter *writer = GEOSWKBWriter_create_r(context);
-    if (writer == NULL) {
-        return -1;
-    }
 
-    GEOSWKBWriter_setOutputDimension_r(context, writer, 3);
-    GEOSWKBWriter_setIncludeSRID_r(context, writer, 1);
-    wkb = GEOSWKBWriter_write_r(context, writer, self->ptr, &size);
-    GEOSWKBWriter_destroy_r(context, writer);
-    if (wkb == NULL) {
-        return -1;
-    }
+    GEOS_INIT;
+    GEOSWKBWriter *writer = GEOSWKBWriter_create_r(ctx);
+    if (writer == NULL) { errstate = PGERR_GEOS_EXCEPTION; goto finish; }
+
+    GEOSWKBWriter_setOutputDimension_r(ctx, writer, 3);
+    GEOSWKBWriter_setIncludeSRID_r(ctx, writer, 1);
+    wkb = GEOSWKBWriter_write_r(ctx, writer, self->ptr, &size);
+    GEOSWKBWriter_destroy_r(ctx, writer);
+    if (wkb == NULL) { errstate = PGERR_GEOS_EXCEPTION; goto finish; }
     x = PyHash_GetFuncDef()->hash(wkb, size);
     if (x == -1) {
         x = -2;
     } else {
         x ^= 374761393UL;  // to make the result distinct from the actual WKB hash //
     }
-    GEOSFree_r(context, wkb);
-    return x;
+    GEOSFree_r(ctx, wkb);
+
+    finish:
+        GEOS_FINISH;
+        if (errstate == PGERR_SUCCESS) {
+            return x;
+        } else {
+            return -1;
```

Comment:
Is there a reason we return -1 here and NULL in the other type member functions? That's just what python expects for those functions?

Reply:
That's what python expects for exceptions:

https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_hash